src/vulkan-renderer/descriptor.cpp

```diff
@@ -0,0 +1,147 @@
+        throw std::runtime_error("Error: vkCreateDescriptorSetLayout failed for descriptor " + name + " !");
+    }
+
+    // TODO: Assign name using debug makers!
+
+    spdlog::debug("Created descriptor sets for descriptor {} successfully.", name);
+}
+
+void Descriptor::reset(const bool clear_descriptor_layout_bindings) {
+    vkDestroyDescriptorSetLayout(device, descriptor_set_layout, nullptr);
+    descriptor_set_layout = VK_NULL_HANDLE;
+
+    vkDestroyDescriptorPool(device, descriptor_pool, nullptr);
+    descriptor_pool = VK_NULL_HANDLE;
+
+    // Only clear descriptor layout bindings when application is shut down.
+    // Do not clear descriptor layout bindings when swapchain needs to be recreated.
+    if (clear_descriptor_layout_bindings) {
+        descriptor_set_layout_bindings.clear();
+    }
+}
+
+void Descriptor::add_descriptor_writes(const std::vector<VkWriteDescriptorSet> &descriptor_writes) {
+    assert(device);
+    assert(descriptor_pool);
+    assert(!name.empty());
+    assert(!descriptor_writes.empty());
+
+    // We need a descriptor write for every descriptor set layout binding!
+    assert(descriptor_set_layout_bindings.size() == descriptor_writes.size());
+
+    this->write_descriptor_sets = descriptor_writes;
+}
+
+void Descriptor::create_descriptor_sets() {
+    assert(device);
+    assert(descriptor_pool);
+    assert(!name.empty());
+    assert(!descriptor_set_layout_bindings.empty());
+    assert(!write_descriptor_sets.empty());
+
+    // We need a descriptor write for every descriptor set layout binding!
+    assert(descriptor_set_layout_bindings.size() == write_descriptor_sets.size());
+
+    spdlog::debug("Creating descriptor sets for '{}'.", name);
+
+    const std::vector<VkDescriptorSetLayout> descriptor_set_layouts(number_of_images_in_swapchain, descriptor_set_layout);
+
+    VkDescriptorSetAllocateInfo descriptor_set_alloc_info = {};
+
+    descriptor_set_alloc_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO;
+    descriptor_set_alloc_info.descriptorPool = descriptor_pool;
+    descriptor_set_alloc_info.descriptorSetCount = static_cast<std::uint32_t>(number_of_images_in_swapchain);
+    descriptor_set_alloc_info.pSetLayouts = descriptor_set_layouts.data();
+
+    // TODO: Do we need this for recreation of swapchain?
+    descriptor_sets.clear();
+
+    descriptor_sets.resize(number_of_images_in_swapchain);
+
+    if (vkAllocateDescriptorSets(device, &descriptor_set_alloc_info, descriptor_sets.data()) != VK_SUCCESS) {
+        throw std::runtime_error("Error: vkAllocateDescriptorSets failed for descriptor " + name + " !");
+    }
+
+    // TODO: Assign name using debug makers!
+
+    for (std::size_t i = 0; i < number_of_images_in_swapchain; i++) {
+        spdlog::debug("Updating descriptor set '{}' #{}", name, i);
+
+        for (std::size_t j = 0; j < write_descriptor_sets.size(); j++) {
```

Comment:
Avoid using `i` and `j` in nested loops, use `k` instead of `j`

Reply:
ok